gusto/transport_equation.py

```diff
@@ -1,13 +1,101 @@
-                       outer, sign, cross, CellNormal, sqrt, Constant,
+                       ds_v, ds_t, ds_b, VectorElement,
+                       outer, sign, cross, CellNormal, Constant,
                        curl, BrokenElement, FunctionSpace)
-from gusto.configuration import logger, DEBUG
+from gusto.configuration import logger, DEBUG, SUPGOptions
 
 
-__all__ = ["LinearAdvection", "AdvectionEquation", "EmbeddedDGAdvection", "SUPGAdvection", "VectorInvariant", "EulerPoincare"]
+__all__ = ["LinearAdvection", "AdvectionEquation", "EmbeddedDGAdvection", "SUPGAdvection", "VectorInvariant", "EulerPoincare", "IntegrateByParts"]
+
+
+class IntegrateByParts(Enum):
+    NEVER = 0
+    ONCE = 1
+    TWICE = 2
+
+
+def is_cg(V):
+    """
+    Function to check is a given space, V, is CG. Broken elements are
+    always discontinuous; for vector elements we check the names of
+    the sobolev spaces of the subelements and for all other elements
+    we just check the sobolev space name.
+    """
+    ele = V.ufl_element()
+    if isinstance(ele, BrokenElement):
+        return False
+    elif type(ele) == VectorElement:
+        return all([e.sobolev_space().name == "H1" for e in ele._sub_elements])
+    else:
+        return V.ufl_element().sobolev_space().name == "H1"
+
+
+def is_dg(V):
+    """
+    Function to check is a given space, V, is DG. Broken elements are
+    always discontinuous; for vector elements we check the names of
+    the sobolev spaces of the subelements and for all other elements
+    we just check the sobolev space name.
+    """
+    ele = V.ufl_element()
+    if isinstance(ele, BrokenElement):
+        return True
+    elif type(ele) == VectorElement:
+        return all([e.sobolev_space().name == "L2" for e in ele._sub_elements])
+    else:
+        return V.ufl_element().sobolev_space().name == "L2"
+
+
+def surface_measures(V):
+    """
+    Function returning the correct surface measures to use for the
+    given function space, V, based on its continuity and also on
+    whether the underlying mesh is extruded.
+    """
+    # CG spaces don't require surface terms
+    if is_cg(V):
+        return None
+    else:
+        # for extruded meshes we have to unpick things
+        if V.extruded:
+            dim = V.mesh().topological_dimension()
+            ele = V.ufl_element()
+
+            # Broken elements are discontinuous so need dS_v and dS_h
+            if type(ele) == BrokenElement:
+                return (dS_v + dS_h)
+
+            # Figure out which spaces we have
+            elif type(ele) == VectorElement:
+                spaces = [ele._sub_elements[i].sobolev_space().name
+                          for i in range(dim)]
+            else:
+                spaces = [ele.sobolev_space()[i].name for i in range(dim)]
+
+            # Based on the space names, these ones are discontinuous
+            discont = [i for i, name in enumerate(spaces) if name in ["L2", "HDiv"]]
```

Comment:
Question: should HDiv be treated as discontinuous? It's not as discontinuous as L2, but not as continuous as H1. I guess for the purpose of this function, you still have surface terms anyways?

Reply:
Yes, for the purpose of this function we need the surface terms. Maybe I should change the comment to "these ones are discontinuous enough to require surface terms"?